mockkubeapiserver/storage/memorystorage/memorystorage_watch.go

```diff
@@ -0,0 +1,119 @@
+package memorystorage
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/kubebuilder-declarative-pattern/mockkubeapiserver/storage"
+)
+
+func (resource *memoryResourceInfo) Watch(ctx context.Context, opt storage.WatchOptions, callback storage.WatchCallback) error {
+	return resource.storage.watch(ctx, resource.gvk, opt, callback)
+}
+
+type resourceStorage struct {
+	parent        *MemoryStorage
+	GroupResource schema.GroupResource
+
+	mutex   sync.Mutex
+	watches []*watch
+
+	resourceVersionClock *resourceVersionClock
+
+	objects map[types.NamespacedName]*unstructured.Unstructured
+}
+
+type resourceVersionClock struct {
+	now atomic.Int64
+}
+
+func (c *resourceVersionClock) Now() int64 {
+	return c.now.Load()
+}
+
+func (c *resourceVersionClock) GetNext() int64 {
+	return c.now.Add(1)
+}
+
+type watch struct {
+	callback storage.WatchCallback
+	opt      storage.WatchOptions
+	errChan  chan error
+}
+
+func (r *resourceStorage) watch(ctx context.Context, gvk schema.GroupVersionKind, opt storage.WatchOptions, callback storage.WatchCallback) error {
+	w := &watch{
+		callback: callback,
+		opt:      opt,
+		errChan:  make(chan error),
+	}
+
+	r.mutex.Lock()
+	pos := -1
+	for i := range r.watches {
+		if r.watches[i] == nil {
+			r.watches[i] = w
+			pos = i
+			break
+		}
+	}
+	if pos == -1 {
+		r.watches = append(r.watches, w)
+		pos = len(r.watches) - 1
+	}
+	r.mutex.Unlock()
+
+	// TODO: Delay / buffer watch notifications until after the list
+
+	// TODO: Only send list if no rv specified?
+
+	r.mutex.Lock()
+	for _, obj := range r.objects {
+		if opt.Namespace != "" {
+			if obj.GetNamespace() != opt.Namespace {
+				continue
+			}
+		}
+
+		ev := storage.BuildWatchEvent(gvk, "ADDED", obj)
```

Comment:
Just to make sure I understand the code correctly: this "ADDED" is a no-op and here we mainly use the watchCallback (next line), right?

Reply:
This is the "real" watch code, the hooks are a sort of "fast-path built-in watch".  When you start a watch with kube-apiserver with resourceVersion unspecified, you get a synthetic ADDED event for every object that exists at that time.

It's actually a streaming LIST (which is otherwise missing from kube), and a little more efficient on the client (at least) in terms of memory, because you only have to process one object at a time, instead of a whole page of objects.  The downside is that it's hard to tell when the initial LIST is over, and you have all the objects.  In the past few versions of kube-apiserver, they added a bookmark notification when the initial list was completed, which is how you can tell.  Maybe I should add bookmark notifications :-)  More info here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/README.md